guacamole/src/main/frontend/src/app/import/controllers/importConnectionsController.js

```diff
@@ -0,0 +1,760 @@
+            // so they can fix their problems and try again.
+            .catch(error => {
+                cleanUpConnections(connectionResponse);
+                handleError(error);
+            }))
+
+        // If an error occured when the call to create the connections was made,
+        // skip any further processing - the user will have a chance to fix the
+        // problems and try again
+        .catch(patchFailure => {
+            $scope.processing = false;
+            $scope.patchFailure = patchFailure;
+        });
+    }
+
+    /**
+     * Display the provided error to the user in a dismissable dialog.
+     *
+     * @argument {ParseError|Error} error
+     *     The error to display.
+     */
+    const handleError = error => {
+
+        // Any error indicates that processing of the file has failed, so clear
+        // all upload state to allow for a fresh retry
+        resetUploadState();
+
+        let text;
+
+        // If it's a import file parsing error
+        if (error instanceof ParseError)
+            text = {
+
+                // Use the translation key if available
+                key: error.key || error.message,
+                variables: error.variables
+            };
+
+        // If it's a generic REST error
+        else if (error instanceof Error)
+            text = error.translatableMessage;
+
+        // If it's an unknown type, just use the message directly
+        else
+            text = { key: error };
+
+        guacNotification.showStatus({
+            className  : 'error',
+            title      : 'IMPORT.DIALOG_HEADER_ERROR',
+            text,
+
+            // Add a button to hide the error
+            actions    : [{
+                name      : 'IMPORT.ACTION_ACKNOWLEDGE',
+                callback  : () => guacNotification.showStatus(false)
+            }]
+        })
+
+    };
+
+    /**
+     * Process the uploaded import file, importing the connections, granting
+     * connection permissions, or displaying errors to the user if there are
+     * problems with the provided file.
+     *
+     * @param {String} mimeType
+     *     The MIME type of the uploaded data file.
+     *
+     * @param {String} data
+     *     The raw string contents of the import file.
+     */
+    function processData(mimeType, data) {
+
+        // Data processing has begun
+        $scope.processing = true;
+
+        // The function that will process all the raw data and return a list of
+        // patches to be submitted to the API
+        let processDataCallback;
+
+        // Choose the appropriate parse function based on the mimetype
+        if (mimeType.endsWith("json"))
+            processDataCallback = connectionParseService.parseJSON;
+
+        else if (mimeType.endsWith("csv"))
+            processDataCallback = connectionParseService.parseCSV;
+
+        else if (mimeType.endsWith("yaml"))
+            processDataCallback = connectionParseService.parseYAML;
```

Comment:
I think we should map the specific, known mimetypes (`application/json`, `text/csv`, etc.) to their corresponding handlers rather than forgivingly accepting anything that ends with specific substrings.

Reply:
Sure, seems fair.